rsocket-core/src/main/java/io/rsocket/RSocketFactory.java

```diff
@@ -0,0 +1,339 @@
+
+import java.time.Duration;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * Creates an RSocket.
+ *
+ * @author Robert Roeser
+ */
+public interface RSocketFactory {
+    /**
+     * Creates a factory that creates RSockets that establish connections to other RSockets
+     */
+    static ClientRSocketFactory connect() {
+        return new ClientRSocketFactory();
+    }
+
+    /**
+     * Creates a factory that creates RSockets that receive connections from other RSockets
+     */
+    static ServerRSocketFactory receive() {
+        return new ServerRSocketFactory();
+    }
+
+    interface Start<T extends Closeable> {
+        Mono<T> start();
+    }
+
+    interface SetupPayload<T> {
+        T setupPayload(Payload payload);
+    }
+
+    interface Transport<T extends io.rsocket.transport.Transport, B extends Closeable> {
+        Start<B> transport(Supplier<T> t);
+
+        default Start<B> transport(T t) {
+            return transport(() -> t);
+        }
+    }
+
+    interface Acceptor<T extends io.rsocket.transport.Transport, A, B extends Closeable> {
+        Transport<T, B> acceptor(Supplier<A> acceptor);
+
+        default Transport<T, B> acceptor(A acceptor) {
+            return acceptor(() -> acceptor);
+        }
+    }
+
+    interface Fragmentation<R extends Acceptor<T, A, B>, T extends io.rsocket.transport.Transport, A, B extends Closeable> {
+        R fragment(int mtu);
+    }
+
+    interface ErrorConsumer<R extends Acceptor<T, A, B>, T extends io.rsocket.transport.Transport, A, B extends Closeable> {
+        R errorConsumer(Consumer<Throwable> errorConsumer);
+    }
+
+    interface KeepAlive<T> {
+        T keepAlive();
+
+        T keepAlive(Duration tickPeriod, Duration ackTimeout, int missedAcks);
+
+        T keepAliveTickPeriod(Duration tickPeriod);
+
+        T keepAliveAckTimeout(Duration ackTimeout);
+
+        T keepAliveMissedAcks(int missedAcks);
+    }
+
+    interface MimeType<T> {
+        T mimeType(String dataMineType, String metadataMimeType);
+
+        T dataMineType(String dataMineType);
+
+        T metadataMimeType(String metadataMimeType);
+    }
+
+    class ClientRSocketFactory implements
+        KeepAlive<ClientRSocketFactory>,
+        MimeType<ClientRSocketFactory>,
+        Acceptor<ClientTransport, RSocket, RSocket>,
+        Transport<ClientTransport, RSocket>,
+        Fragmentation<ClientRSocketFactory, ClientTransport, RSocket, RSocket>,
+        ErrorConsumer<ClientRSocketFactory, ClientTransport, RSocket, RSocket>,
+        SetupPayload<ClientRSocketFactory> {
+
+        private Supplier<RSocket> acceptor = () -> new AbstractRSocket() {
+        };
+        private Supplier<io.rsocket.transport.ClientTransport> transportClient;
+        private Consumer<Throwable> errorConsumer = Throwable::printStackTrace;
+        private int mtu = 0;
+        private int flags = SetupFrameFlyweight.VALID_FLAGS - SetupFrameFlyweight.FLAGS_RESUME_ENABLE;
```

Comment:
@yschimke I removed leasing for the moment - trying to think of a way to add that back in cleaner manner than before. Should have added everything is disabled by default now, leasing, fragmentation, keep-alive, etc. You have to explicitly enable them.